src/github.com/getlantern/autoupdate-server/server/server.go

```diff
@@ -0,0 +1,153 @@
+	// hardware architecture of target platform
+	Arch string `json:"-"`
+	// application-level user identifier
+	//UserId string `json:"user_id"`
+	// checksum of the binary to replace (used for returning diff patches)
+	Checksum string `json:"checksum"`
+	// release channel (empty string means 'stable')
+	//Channel string `json:"-"`
+	// tags for custom update channels
+	Tags map[string]string `json:"tags"`
+}
+
+// Result represents the answer to be sent to the client.
+type Result struct {
+	// should the update be applied automatically/manually
+	Initiative Initiative `json:"initiative"`
+	// url where to download the updated application
+	URL string `json:"url"`
+	// a URL to a patch to apply
+	PatchURL string `json:"patch_url"`
+	// the patch format (only bsdiff supported at the moment)
+	PatchType PatchType `json:"patch_type"`
+	// version of the new application
+	Version string `json:"version"`
+	// expected checksum of the new application
+	Checksum string `json:"checksum"`
+	// signature for verifying update authenticity
+	Signature string `json:"signature"`
+}
+
+// CheckForUpdate receives a *Params message and emits a *Result. If both res
+// and err are nil it means no update is available.
+func (g *ReleaseManager) CheckForUpdate(p *Params) (res *Result, err error) {
+
+	// Keep for the future.
+	if p.Version < 1 {
+		p.Version = 1
+	}
+
+	// p must not be nil.
+	if p == nil {
+		return nil, fmt.Errorf("Expecting params")
+	}
+
+	if p.Tags != nil {
+		// Compatibility with go-check.
+		if p.Tags["os"] != "" {
+			p.OS = p.Tags["os"]
+		}
+		if p.Tags["arch"] != "" {
+			p.Arch = p.Tags["arch"]
+		}
+	}
+
+	if !isVersionTag(p.AppVersion) {
+		return nil, fmt.Errorf("Expecting a version tag of the form vX.Y.Z.")
+	}
+
+	if p.Checksum == "" {
+		return nil, fmt.Errorf("Checksum must not be nil")
+	}
+
+	if p.OS == "" {
+		return nil, fmt.Errorf("OS is required")
+	}
+
+	if p.Arch == "" {
+		return nil, fmt.Errorf("Arch is required")
+	}
+
+	// Looking if there is a newer version for the os/arch.
+	var update *Asset
+	if update, err = g.getProductUpdate(p.OS, p.Arch); err != nil {
+		return nil, fmt.Errorf("Could not lookup for updates: %s", err)
+	}
+
+	// Looking for the asset thay matches the current app checksum.
+	var current *Asset
+	if current, err = g.lookupAssetWithChecksum(p.OS, p.Arch, p.Checksum); err != nil {
+		// No such asset with the given checksum, nothing to compare.
+
+		r := &Result{
+			Initiative: INITIATIVE_AUTO,
+			URL:        assetURL(update.URL),
+			PatchType:  PATCHTYPE_NONE,
+			Version:    update.v,
+			Checksum:   update.Checksum,
+			Signature:  update.Signature,
+		}
+
+		return r, nil
+	}
+
+	// No update available.
+	if VersionCompare(p.AppVersion, update.v) != Higher {
+		return nil, ErrNoUpdateAvailable
+	}
+
+	// A newer version is available!
+
+	// Generate a binary diff of the two assets.
+	var patch *Patch
+	if patch, err = GeneratePatch(current.URL, update.URL); err != nil {
```

Comment:
It looks like this is getting called for every request.  That seems like it would get expensive.